[jit] allow RRef local creation with IValue objects

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#33263 [jit] allow RRef local creation with IValue objects**

This PR allow PyRRef local creation to inspect the pyobject, if it
founds that we could turn it to an IValue, turn to an IValue first,
otherwise hold it as a PyObjectType

One part is missing: python_cu might be different across workers, so
need to make a type_hint based approach for script module and module
interface. That will be separately done.

Differential Revision: [D19871243](https://our.internmc.facebook.com/intern/diff/D19871243)

wanchaol committed:
[jit] allow RRef local creation with IValue objects

This PR allow PyRRef local creation to inspect the pyobject, if it
founds that we could turn it to an IValue, turn to an IValue first,
otherwise hold it as a PyObjectType

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] allow RRef local creation with IValue objects"


This PR allow PyRRef local creation to inspect the pyobject, if it
founds that we could turn it to an IValue, turn to an IValue first,
otherwise hold it as a PyObjectType

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] allow RRef local creation with IValue objects"


This PR allow PyRRef local creation to inspect the pyobject, if it
founds that we could turn it to an IValue, turn to an IValue first,
otherwise hold it as a PyObjectType

Differential Revision: [D19871243](https://our.internmc.facebook.com/intern/diff/D19871243)

[ghstack-poisoned]

zhaojuanmao commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -1745,3 +1752,19 @@ def rref_tensor_is_owner(rref_var):
+    def test_local_rref_with_script_module(self):
+        # create a local RRef that hold a ScriptModule
+        script_mod = MyScriptModule()
+        rref_module = rpc.RRef(script_mod._c)
```
How about using rpc_sync call here?  rref_module = rpc_sync(workName, create_local_rref, args=())

def create_local_rref:
     return rpc.RRef(script_mod._c)

wanchaol replied:
what's the reason we are wrapping this with `rpc_sync` call?

zhaojuanmao replied:
I think this is the use case we want? otherwise how the rref_module is passed?

rref_module = rpc_sync(workName, create_local_rref, args=())
rpc_sync(workName, run_rref_forward, args=(rref_module, ...))

wanchaol replied:
sounds good I will add this use case.

wanchaol committed:
Update on "[rpc] allow RRef local creation with IValue objects"


This PR allow PyRRef local creation to inspect the pyobject, if it
founds that we could turn it to an IValue, turn to an IValue first,
otherwise hold it as a PyObjectType

Differential Revision: [D19871243](https://our.internmc.facebook.com/intern/diff/D19871243)

[ghstack-poisoned]

wanchaol committed:
Update on "[rpc] allow RRef local creation with IValue objects"


This PR allow PyRRef local creation to inspect the pyobject, if it
founds that we could turn it to an IValue, turn to an IValue first,
otherwise hold it as a PyObjectType

Differential Revision: [D19871243](https://our.internmc.facebook.com/intern/diff/D19871243)

[ghstack-poisoned]

wanchaol committed:
Update on "[rpc] allow RRef local creation with IValue objects"


This PR allow PyRRef local creation to inspect the pyobject, if it
founds that we could turn it to an IValue, turn to an IValue first,
otherwise hold it as a PyObjectType

Differential Revision: [D19871243](https://our.internmc.facebook.com/intern/diff/D19871243)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/py_rref.cpp:
```diff
@@ -65,12 +65,24 @@ PyRRef::PyRRef(c10::intrusive_ptr<RRef> rref) : rref_(std::move(rref)) {
+        IValue ivalue = jit::toIValue(std::move(copy), elem_type);
+        rref->setValue(std::move(ivalue));
         return rref;
+
```
remove empty line?

mrshenli commented on torch/csrc/distributed/rpc/py_rref.cpp:
```diff
@@ -65,12 +65,24 @@ PyRRef::PyRRef(c10::intrusive_ptr<RRef> rref) : rref_(std::move(rref)) {
+          elem_type = type_inferred.type();
+        } else {
+          // Otherwise it's a pure pyobject, create the RRef
+          // that holds a IValue of an pyobject
```
**an** IValue of **a** pyobject

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -1745,3 +1752,19 @@ def rref_tensor_is_owner(rref_var):
+
+    @dist_init
+    def test_local_rref_with_script_module(self):
+        # create a local RRef that hold a ScriptModule
```
holds

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -1745,3 +1752,19 @@ def rref_tensor_is_owner(rref_var):
+            return rref_module.to_here().forward()
+
+        # pass the local created RRef module to jit
+        res = rref_script_module(rref_module)
```
this is pure local test? can we pass this RRef to remote workers?

wanchaol replied:
This right now could not be pass RRef to remote workers, it needs to work upon the jit pickler https://github.com/pytorch/pytorch/pull/32959, I will first test this locally as it is enough to distinguish from local test that we are holding a `ScriptModule` type in the RRef instead of a `PyObjectType`.

wanchaol replied:
hmm seems like even a local test with a remote rref need jit pickler/unpickler change. I will wait for that PR to land first and then land this.

mrshenli commented on torch/csrc/distributed/rpc/py_rref.cpp:
```diff
@@ -65,12 +65,24 @@ PyRRef::PyRRef(c10::intrusive_ptr<RRef> rref) : rref_(std::move(rref)) {
+        if (type_inferred.success()) {
+          // If we could infer the type from the pyobject, we create
+          // the RRef with the IValue of that type.
+          elem_type = type_inferred.type();
```
Is there a way to test this? Say I do `RRef(1)`, then the `RRef` should beholding an IValue type int instead of `py::object`.

wanchaol replied:
I did test this through the module interface, a local created RRef (without passing a torchscript function or sth) is now associated with the JIT type instead of a blind `PyObjectType`, so it's compatible with the module interface type we defined in the type comment. we can also pass a tensor to it and call some tensor methods afterward, I will add more tests cases once https://github.com/pytorch/pytorch/pull/32959 landed.

mrshenli reviewed (commented):
LGTM! Thanks! Left some comments inline.

zhaojuanmao commented:
do you want to patch jit pickler diff and run test and see whether it works fine?

wanchaol commented:
> do you want to patch jit pickler diff and run test and see whether it works fine?

sure I will try that.

wanchaol committed:
Update on "[rpc] allow RRef local creation with IValue objects"


This PR allow PyRRef local creation to inspect the pyobject, if it
founds that we could turn it to an IValue, turn to an IValue first,
otherwise hold it as a PyObjectType

Differential Revision: [D19871243](https://our.internmc.facebook.com/intern/diff/D19871243)

[ghstack-poisoned]

wanchaol committed:
Update on "[rpc] allow RRef local creation with IValue objects"


This PR allow PyRRef local creation to inspect the pyobject, if it
founds that we could turn it to an IValue, turn to an IValue first,
otherwise hold it as a PyObjectType

Differential Revision: [D19871243](https://our.internmc.facebook.com/intern/diff/D19871243)

[ghstack-poisoned]

wanchaol committed:
Update on "[rpc] allow RRef local creation with IValue objects"


This PR allow PyRRef local creation to inspect the pyobject, if it
founds that we could turn it to an IValue, turn to an IValue first,
otherwise hold it as a PyObjectType

Differential Revision: [D19871243](https://our.internmc.facebook.com/intern/diff/D19871243)

[ghstack-poisoned]

wanchaol committed:
Update on "[rpc] allow RRef local creation with IValue objects"


This PR allow PyRRef local creation to inspect the pyobject, if it
founds that we could turn it to an IValue, turn to an IValue first,
otherwise hold it as a PyObjectType

Differential Revision: [D19871243](https://our.internmc.facebook.com/intern/diff/D19871243)

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] allow RRef local creation with IValue objects"


This PR allow PyRRef local creation to inspect the pyobject, if it
founds that we could turn it to an IValue, turn to an IValue first,
otherwise hold it as a PyObjectType

Differential Revision: [D19871243](https://our.internmc.facebook.com/intern/diff/D19871243)

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] allow RRef local creation with IValue objects"


This PR allow PyRRef local creation to inspect the pyobject, if it
founds that we could turn it to an IValue, turn to an IValue first,
otherwise hold it as a PyObjectType

Differential Revision: [D19871243](https://our.internmc.facebook.com/intern/diff/D19871243)

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] allow RRef local creation with IValue objects"


This PR allow PyRRef local creation to inspect the pyobject, if it
founds that we could turn it to an IValue, turn to an IValue first,
otherwise hold it as a PyObjectType

Differential Revision: [D19871243](https://our.internmc.facebook.com/intern/diff/D19871243)

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] allow RRef local creation with IValue objects"


This PR allow PyRRef local creation to inspect the pyobject, if it
founds that we could turn it to an IValue, turn to an IValue first,
otherwise hold it as a PyObjectType

Differential Revision: [D19871243](https://our.internmc.facebook.com/intern/diff/D19871243)

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] allow RRef local creation with IValue objects"


This PR allow PyRRef local creation to inspect the pyobject, if it
founds that we could turn it to an IValue, turn to an IValue first,
otherwise hold it as a PyObjectType

Differential Revision: [D19871243](https://our.internmc.facebook.com/intern/diff/D19871243)

[ghstack-poisoned]

wanchaol committed:
Update on "[jit] allow RRef local creation with IValue objects"


This PR allow PyRRef local creation to inspect the pyobject, if it
founds that we could turn it to an IValue, turn to an IValue first,
otherwise hold it as a PyObjectType

One part is missing: python_cu might be different across workers, so
need to make a type_hint based approach for script module and module
interface. That will be separately done.

Differential Revision: [D19871243](https://our.internmc.facebook.com/intern/diff/D19871243)

[ghstack-poisoned]

wanchaol commented:
@zhaojuanmao @mrshenli enabling passing RRefs to other workers with scriptmodule and module interface will be done separately by @xush6528, this patch will solely be just: local rref creation with jit types.

wanchaol committed:
Update on "[jit] allow RRef local creation with IValue objects"


This PR allow PyRRef local creation to inspect the pyobject, if it
founds that we could turn it to an IValue, turn to an IValue first,
otherwise hold it as a PyObjectType

One part is missing: python_cu might be different across workers, so
need to make a type_hint based approach for script module and module
interface. That will be separately done.

Differential Revision: [D19871243](https://our.internmc.facebook.com/intern/diff/D19871243)

[ghstack-poisoned]

zhaojuanmao commented on torch/testing/_internal/distributed/rpc/jit/rpc_test.py:
```diff
@@ -249,3 +249,14 @@ def rref_script_annotation(rref_var):
+        self.assertEqual(rref_local_script_class.to_here().a, 10)
+
+        # create a local RRef that holds a ScriptModule
+        rref_local_script_mod = rpc.RRef(MyScriptModule(3)._c)
```
So users have to write "._c" to construct script module, right?

wanchaol replied:
yeah, this will give the jit_type of the script module locally. We can infer the scriptModule `py::object` directly to get its type, but I think that overkills the purpose of this PR. I will leave that with separate PR that enabling complete ScriptModule and ModuleInterface rref creation across workers. cc @xush6528

wanchaol committed:
Update on "[jit] allow RRef local creation with IValue objects"


This PR allow PyRRef local creation to inspect the pyobject, if it
founds that we could turn it to an IValue, turn to an IValue first,
otherwise hold it as a PyObjectType

One part is missing: python_cu might be different across workers, so
need to make a type_hint based approach for script module and module
interface. That will be separately done.

Differential Revision: [D19871243](https://our.internmc.facebook.com/intern/diff/D19871243)

[ghstack-poisoned]
